uaclient/entitlements/repo.py

```diff
@@ -95,11 +95,90 @@ def _perform_enable(self, silent: bool = False) -> bool:
         return True
 
     def _perform_disable(self, silent=False):
+        if self.purge:
+            repo_origin_packages = apt.get_installed_packages_by_origin(
+                # this is for mypy, at this point we know origin exists
+                self.origin
+                or ""
+            )
+            packages_to_reinstall = []
+            packages_to_remove = []
+            for package in repo_origin_packages:
+                alternatives = apt.get_alternative_versions_for_package(
+                    package, exclude_origin=self.origin
+                )
+                if alternatives:
+                    # We can call max(List[Version]) but mypy doesn't know.
+                    # Or doesn't like it.
+                    packages_to_reinstall.append(
+                        (package, max(alternatives))  # type: ignore
+                    )
+                else:
+                    packages_to_remove.append(package)
+
+            if not self.prompt_for_purge(
+                packages_to_remove, packages_to_reinstall
+            ):
+                return False
+
         if hasattr(self, "remove_packages"):
             self.remove_packages()
         self.remove_apt_config(silent=silent)
+
+        if self.purge:
+            self.execute_purge(packages_to_remove, packages_to_reinstall)
         return True
 
+    def prompt_for_purge(self, packages_to_remove, packages_to_reinstall):
+        prompt = False
+        if packages_to_remove:
+            print(messages.WARN_PACKAGES_REMOVAL)
+            util.print_package_list(
+                [package.name for package in packages_to_remove]
+            )
+            prompt = True
+
+        if packages_to_reinstall:
+            print(messages.WARN_PACKAGES_REINSTALL)
+            util.print_package_list(
+                [package.name for (package, _) in packages_to_reinstall]
+            )
+            prompt = True
+
+        if prompt:
+            return util.prompt_for_confirmation(messages.PROCEED_YES_NO)
+        return True
+
+    def execute_purge(self, packages_to_remove, packages_to_reinstall):
+        to_remove = [
+            package.name
+            for package in packages_to_remove
+            if package.current_ver
```

Comment:
Can't we guarantee the the `package.current_ver` is always set when creating the `packages_to_remove` list ?

Reply:
yes we can, because we only pass installed packages to the function which returns those. However, some entitlements have this `packages_to_remove` directive/variable, and then some of those packages can be removed in between the purge-driven selection of packages and the actual purge - `ubuntu-fips` is a good example when purging FIPS.

Reply:
I think that nuance is worth adding here as a comment, otherwise it may be tempting to some future developer to remove the check